File: tests/Dle.UnitTests/Attribution/ClaimCodeTests.cs
using Dle.Domain.Attribution;
using Xunit;

namespace Dle.UnitTests.Attribution;

/// <summary>
/// FR-184 and TC-148. The claim code is read off an interstitial by a human and typed into an
/// application, so the alphabet deliberately omits every confusable glyph. "O" must not be silently
/// mapped onto "0": the code the user reads never contains either, and a quiet mapping would turn a
/// typo into someone else's attribution.
/// </summary>
public sealed class ClaimCodeTests
{
    [Fact]
    public void Alphabet_OmitsEveryConfusableGlyph()
    {
        Assert.Equal("ACDEFGHJKLMNPQRTUVWXY34679", ClaimCode.Alphabet);

        foreach (char confusable in "BIOSZ01258")
        {
            Assert.DoesNotContain(confusable, ClaimCode.Alphabet);
        }
    }

    [Fact]
    public void Length_IsSix()
    {
        Assert.Equal(6, ClaimCode.Length);
    }

    [Theory]
    [InlineData("acdefg", "ACDEFG")]
    [InlineData("ACD-EFG", "ACDEFG")]
    [InlineData(" ACD EFG ", "ACDEFG")]
    [InlineData("acd-efg", "ACDEFG")]
    [InlineData("A C-D E F G", "ACDEFG")]
    public void Normalize_UppercasesAndStripsSeparators(string raw, string expected)
    {
        Assert.Equal(expected, ClaimCode.Normalize(raw));
    }

    [Fact]
    public void Normalize_DoesNotMapOToZero()
    {
        // "O" is not in the alphabet, so a code containing it is a typo and must fail validation
        // rather than be repaired into a different, possibly valid, code.
        string normalized = ClaimCode.Normalize("ACDEFO");

        Assert.Equal("ACDEFO", normalized);
        Assert.False(ClaimCode.IsWellFormed(normalized));
    }

    [Fact]
    public void Normalize_EmptyInput_IsEmpty()
    {
        Assert.Equal(string.Empty, ClaimCode.Normalize(string.Empty));
        Assert.Equal(string.Empty, ClaimCode.Normalize("   "));
        Assert.Equal(string.Empty, ClaimCode.Normalize("---"));
    }

    [Fact]
    public void Normalize_NullInput_Throws()
    {
        Assert.Throws<ArgumentNullException>(() => ClaimCode.Normalize(null!));
    }

    [Fact]
    public void Normalize_IsIdempotent()
    {
        string once = ClaimCode.Normalize("acd-efg");

        Assert.Equal(once, ClaimCode.Normalize(once));
    }

    [Theory]
    [InlineData("ACDEFG", true)]
    [InlineData("34679A", true)]
    [InlineData("XYWVUT", true)]
    [InlineData("ABCDEF", false)]   // B is not in the alphabet
    [InlineData("ACDEF0", false)]   // zero is not in the alphabet
    [InlineData("ACDEF1", false)]
    [InlineData("ACDEF8", false)]
    [InlineData("acdefg", false)]   // must be normalised first
    [InlineData("ACDEF", false)]    // too short
    [InlineData("ACDEFGH", false)]  // too long
    [InlineData("", false)]
    [InlineData(null, false)]
    public void IsWellFormed_ChecksLengthAndAlphabet(string? code, bool expected)
    {
        Assert.Equal(expected, ClaimCode.IsWellFormed(code));
    }
}

File: sdk/ios/Tests/DleSDKTests/ClaimCodeTests.swift
import XCTest
@testable import DleSDK

/// Mirrors `tests/Dle.UnitTests/Attribution/ClaimCodeTests.cs` case for case, so the SDK and
/// the engine agree on what a claim code is.
final class ClaimCodeTests: XCTestCase {
    func testAlphabetAndLengthMatchTheDomain() {
        XCTAssertEqual(DleClaimCode.alphabet, "ACDEFGHJKLMNPQRTUVWXY34679")
        XCTAssertEqual(DleClaimCode.alphabet.count, 26)
        XCTAssertEqual(DleClaimCode.length, 6)
        for homoglyph in "BIOSZ01258" {
            XCTAssertFalse(DleClaimCode.alphabet.contains(homoglyph), "\(homoglyph) must not be in the alphabet")
        }
    }

    func testNormalizeUpperCasesAndStripsSeparators() {
        let cases: [(String, String)] = [
            ("acdefg", "ACDEFG"),
            ("ACD-EFG", "ACDEFG"),
            (" ACD EFG ", "ACDEFG"),
            ("acd-efg", "ACDEFG"),
            ("A C-D E F G", "ACDEFG"),
            ("acd\tefg\n", "ACDEFG"),
            ("acd\u{00A0}efg", "ACDEFG"),
        ]
        for (raw, expected) in cases {
            XCTAssertEqual(DleClaimCode.normalize(raw), expected, "normalize(\(raw.debugDescription))")
        }
    }

    func testNormalizeDoesNotSubstituteCharacters() {
        // O is not rewritten to 0: neither is in the alphabet, and rewriting could turn a typo
        // into a different valid code.
        let normalized = DleClaimCode.normalize("ACDEFO")
        XCTAssertEqual(normalized, "ACDEFO")
        XCTAssertFalse(DleClaimCode.isWellFormed(normalized))
    }

    func testNormalizeOfSeparatorsOnlyIsEmpty() {
        XCTAssertEqual(DleClaimCode.normalize(""), "")
        XCTAssertEqual(DleClaimCode.normalize("   "), "")
        XCTAssertEqual(DleClaimCode.normalize("---"), "")
    }

    func testNormalizeIsIdempotent() {
        let once = DleClaimCode.normalize("acd-efg")
        XCTAssertEqual(once, DleClaimCode.normalize(once))
    }

    func testNormalizeKeepsCharactersWithoutASimpleUpperCase() {
        // `char.ToUpper` maps one UTF-16 unit to one unit; ß has no such mapping and stays.
        XCTAssertEqual(DleClaimCode.normalize("aß"), "Aß")
    }

    func testIsWellFormed() {
        let cases: [(String?, Bool)] = [
            ("ACDEFG", true),
            ("34679A", true),
            ("XYWVUT", true),
            ("ABCDEF", false), // B is not in the alphabet
            ("ACDEF0", false), // zero is not in the alphabet
            ("ACDEF1", false),
            ("ACDEF8", false),
            ("acdefg", false), // must be normalised first
            ("ACDEF", false), // too short
            ("ACDEFGH", false), // too long
            ("", false),
            (nil, false),
            ("ACDÉFG", false),
        ]
        for (code, expected) in cases {
            XCTAssertEqual(DleClaimCode.isWellFormed(code), expected, "isWellFormed(\(String(describing: code)))")
        }
    }

    func testEveryAlphabetCharacterFormsAValidCode() {
        for character in DleClaimCode.alphabet {
            XCTAssertTrue(DleClaimCode.isWellFormed(String(repeating: character, count: 6)))
        }
    }
}
